File: samples/canfd_test.cpp
// Copyright 2025 Reazon Holdings, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// ============================================================================
// Single Motor CANFD Control Test
//
// This program tests controlling a single motor via CAN-FD using MIT control mode.
// It continuously sends a constant torque command to the motor at a fixed control rate.
//
// Features:
// - Uses MIT mode to send a torque command of 1.4 Nm.
// - Reads and prints motor position, velocity, and torque in real-time.
//
// This is useful for verifying basic communication and control behavior
// over CAN-FD with a single motor before expanding to multi-motor systems.
// ============================================================================

#include <atomic>
#include <string>
#include <chrono>
#include <csignal>  
#include <vector>
#include <iostream>
#include <thread>
#include <fstream>
#include <sstream>
#include "../src/controller/dynamics.hpp"
#include "../src/dmmotor/damiao_port.hpp"

#define FOLLOWER_DEVICENAME0 "can0"
#define TICK 0.002
#define DOF 1 
#define PI 3.141592665

std::atomic<bool> running{true};

void signalHandler(int) {
    std::cout << "\n[Signal] SIGINT caught. Shutting down gracefully...\n";
    running = false;  
}

int main(int argc, char **argv){
    std::signal(SIGINT, signalHandler);

    DamiaoPort follower(
        FOLLOWER_DEVICENAME0, 
        {DM_Motor_Type::DM4340},
        {0x02},
        {0x12},
        {true},
        Control_Type::MIT,
        CAN_MODE_FD);

    //follower.setZeroPosition();
    //follower.moveTorqueSync2({0});

    std::vector<double> joint_positions(DOF, 0.0);
    std::vector<double> joint_velocities(DOF, 0.0);
    std::vector<double> joint_torques(DOF, 0.0);
    std::vector<double> command(DOF, 0.0);
    std::cout << "CANFD TEST START!!!" << std::endl;
    while (running) {
        auto loop_start_time = std::chrono::steady_clock::now();

        for (size_t i = 0; i < DOF; ++i) {
            joint_positions[i] = follower.motors_[i]->getPosition();
            joint_velocities[i] = follower.motors_[i]->getVelocity();
            joint_torques[i] = follower.motors_[i]->getTorque();

            //std::cout << "joint_positions[" << i << "] = " << joint_positions[i] << std::endl;
            //std::cout << "joint_velocities[" << i << "] = " << joint_velocities[i] << std::endl;
            //std::cout << "joint_torques[" << i << "] = " << joint_torques[i] << std::endl;
        }

        std::vector<double> kp_temp = {10.0};
        std::vector<double> kd_temp = {0.1};
        std::vector<double> pos_temp = {3.14};
        std::vector<double> vel_temp = {10000};
        command = {10000};

        // setMIT mode
        //follower.setMIT(pos_temp, vel_temp, kp_temp, kd_temp, command);
        
        // setMITSync mode
        //follower.setMITSync(pos_temp, vel_temp, kp_temp, kd_temp, command);

        // setPosVel mode
        // follower.setPosVel(pos_temp, vel_temp);

        // setPosVelSync mode
         //follower.setPosVelSync(pos_temp, vel_temp);

        // setPosForce mode
         //follower.setPosForce(pos_temp, vel_temp, command);

        // setPosForceSync mode
         //follower.setPosForceSync(pos_temp, vel_temp, command);

        // setVel mode
        //follower.setVel(vel_temp);

        // setVelSync mode
       //follower.setVelSync(vel_temp);

        auto loop_end_time = std::chrono::steady_clock::now();
        auto t_elapsed = std::chrono::duration<double>(loop_end_time - loop_start_time).count();
        std::cout << "elapsed time [s] : " << t_elapsed << std::endl;
        std::cout << "control Frequency [Hz]" << 1.0/t_elapsed << std::endl;
        auto sleep_time = std::chrono::duration<double>(TICK) - (loop_end_time - loop_start_time);
        if (sleep_time.count() > 0.0) {
            std::this_thread::sleep_for(sleep_time);
        } else {
            std::cout << "Loop execution time exceeded the tick duration!" << std::endl;
        }
    }

    std::cout << "finished" << std::endl;
    follower.disable();

    return 0;
}

File: script/socketcan_control/canfd_test.py
#!/usr/bin/env python3
#
# Copyright 2025 Reazon Holdings, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from DM_SocketCANFD import *
from time import sleep
import numpy as np
from numpy import pi
import time

CAN_DEVICE_NAME = "can0"
NUMJOINT = 1
Kp = [0]
Kd = [0]
# Pos = [np.pi*2]
Pos = [0]
Vel = [0]
Tau = [1.0]

zeros = np.zeros(NUMJOINT)

if __name__ == "__main__":

    def main():
        openarm = DamiaoPort(
            CAN_DEVICE_NAME,
            [
                DM_Motor_Type.DM4340
            ],
            [0x08],
            [0x18],
            use_canfd = True
        )

        openarm.enable()

        try:
            while True:
                start = time.perf_counter()

                # openarm.set_goal_positions_sync(Pose, Kp, Kd)
                # openarm.setMIT(Kp, Kd, Pos, Vel, Tau)
                # openarm.set_goal_posvel(Pos)
                openarm.setMITSync(Kp, Kd, Pos, Vel, Tau)
                # openarm.setVel(Vel)
                # openarm.setVelSync(Vel)
                # openarm.setPosVel(Pos,Vel)
                # openarm.setPosVelSync(Pos,Vel)
                # openarm.setPosForce(Pos,Vel,Tau)
                # openarm.setPosForceSync(Pos, Vel, Tau)

                status = openarm.get_present_status()

                for i, motor_status in enumerate(status):
                    print(f"[Motor {i}] goal_pos: {motor_status[0]:.3f}, "
                          f"goal_tau: {motor_status[1]:.3f}, "
                          f"pos: {motor_status[2]:.3f}, "
                          f"vel: {motor_status[3]:.3f}, "
                          f"tau: {motor_status[4]:.3f}, "
                          f"tmos: {motor_status[5]}, "
                          f"trotor: {motor_status[6]}")
                
                end = time.perf_counter()
                elapsed = end - start
                freq = 1.0 / elapsed if elapsed > 0 else 0.0

                print(f"Loop time: {elapsed*1000:.3f} ms, Frequency: {freq:.1f} Hz")
                print("="*100)

            

        except KeyboardInterrupt:
            print("Interrupted by user. Disabling motors...")
            openarm.disable()
            print("Shutdown complete.")

    main()

